src/engine-actions/prepare-swap.js

```diff
@@ -23,23 +14,32 @@ const MEMO_PREFIX = 'sparkswap-swap-pivot:'
  * Prepares for a swap in which this node is the counterparty to the intiating node
  *
  * @param {string} orderId - Order for which the swap is being executed
- * @param {string} swapHash        - swap hash that will be associated with the swap
- * @param {string} value   - Int64 string of the value of inbound currency
- * @returns {string} swapHash      Hash that will be translated
+ * @param {string} swapHash - swap hash that will be associated with the swap
+ * @param {string} value - Int64 string of the value of inbound currency
+ * @param {number} expiry - payment request expiry time in seconds
+ * @param {number} cltvExpiry - delta to use for the time-lock of the CLTV
+ *                              extended to the final hop (in blocks)
+ * @returns {string} paymentRequest - lightning invoice for a payment
  */
-async function prepareSwap (orderId, swapHash, value) {
+async function prepareSwap (orderId, swapHash, value, expiry, cltvExpiry) {
   this.logger.info(`Preparing swap for ${swapHash}`, { value })
 
-  const params = {
-    memo: `${MEMO_PREFIX}${orderId}`,
-    value,
-    expiry: SWAP_EXPIRY,
-    hash: swapHash
-  }
-
-  const { paymentRequest } = await addHoldInvoice(params, { client: this.client })
+  // make prepareSwap idempotent by returning the existing invoice if one exists
+  try {
+    const { paymentRequest } = await lookupInvoice(swapHash, { client: this.client })
```

Comment:
does the call to lookupInvoice throw an error or is it the destructuring? I ask because we will create an invoice _for any_ error that occurs, which might have consequences since we only want to add an invoice if it's not found.

Reply:
It looks like `lookupInvoice` throws an error when the invoice is not found (the documentation isn't really clear but I looked at the source). I don't think the destructuring can produce an error here. There could be an error if `swapHash` is invalid, but that should fail in the catch block also.

Reply:
It looks like lookupinvoice will send a code `2` with the message `Error: 2 UNKNOWN: there are no existing invoices`. We should match on this error and only add a new invoice if it's this code.

Reply:
Is that documented?

Reply:
Not documented. This is the functionality when you use a `swapHash` that doesn't exist. We do a similar check here: https://github.com/sparkswap/lnd-engine/blob/77b02eb69327477f123d49b71da865bdfce62809/src/engine-actions/get-status.js#L76

It is semi-fragile. GRPC has its own status codes, the only way to differentiate between the errors is through the message.